Spark 4.0: RewriteTablePath: Update sizes of rewritten manifests in manifest lists

vaultah opened the pull request:
Fixes https://github.com/apache/iceberg/issues/13719

The idea is to rewrite manifest files first, in order to get their final sizes, then rewrite manifest lists updating sizes for the previously rewritten manifests.

dramaticlly commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -415,54 +421,85 @@ private Set<Pair<String, String>> statsFileCopyPlan(
-                .collectAsList());
+        expectedManifestPaths =
+            Sets.newHashSet(
+                allManifestsDF
+                    .distinct()
+                    .filter(
+                        functions
+                            .column(ManifestFile.SNAPSHOT_ID.name())
+                            .isInCollection(deltaSnapshotIds))
+                    .as(Encoders.STRING())
+                    .collectAsList());
       }
+
+      Set<ManifestFile> foundManifests =
+          deltaSnapshots.stream()
+              .flatMap(
+                  s -> {
+                    try {
+                      return s.allManifests(table.io()).stream();
+                    } catch (NotFoundException e) {
+                      LOG.warn(
+                          "Skipping snapshot {} as its manifest list is missing (likely expired).",
+                          s.snapshotId(),
+                          e);
+                      return Stream.empty();
+                    }
+                  })
+              .collect(Collectors.toSet());
+
+      Set<String> foundManifestPaths =
+          foundManifests.stream().map(ManifestFile::path).collect(Collectors.toSet());
+      Set<String> missingPaths = Sets.difference(expectedManifestPaths, foundManifestPaths);
+      Preconditions.checkState(
+          missingPaths.isEmpty(),
+          "Could not find all expected manifests. Missing files: %s",
+          String.join(", ", missingPaths));
+
+      return foundManifests.stream()
+          .filter(m -> expectedManifestPaths.contains(m.path()))
+          .collect(Collectors.toSet());
```
If you are looking for set of manifests filtered by snapshot ids, I think this can probably be replaced with something similar to https://github.com/apache/iceberg/blob/main/spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java#L182-L190,

vaultah replied:
We can get `expectedManifestPaths` using `manifestDF`, but in the end we still need `ManifestFile` objects.

This logic needs to be changed though, because we always have to rewrite all referenced manifest files for the selected snapshots to make sure their size are correct

dramaticlly replied:
yeah I think we might want the same thing? 

From what I can tell,  [`.all_manifests` table](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/AllManifestsTable.java) include all the valid manifest file that are referenced from any snapshot currently tracked by the
table, and if we apply the filter on referenced_snapshot with deltaSnapshotIds, it shall provide us the same result as `foundManifests`. I am thinking if we use ManifestEncoders against returned `Dataset<Row>` and collect as a list (or a set) of manifestsToRewrite

dramaticlly reviewed (commented):
Thanks for the change and I think this is on the right direction. Left some comments and normally I would suggest we focus the changeset to latest Spark version 4.0 and have a separate PR to backfill the changes for older spark version 3.4/3.5 after PR is approved. 

@anuragmantri, last I heard you are interested in this as well

dramaticlly commented:
@szehon-ho or @amogh-jahagirdar can you help kick off the CI ?

anuragmantri commented:
Yes, this is great timing, I just started working on a fix for this. I had a very similar approach in mind. @vaultah do you mind fixing the position delete files rewrite as well?  see https://github.com/apache/iceberg/issues/12554

We generally add changes to only the default spark version (4.0 now) and backport to older versions later. Makes reviews easier. Thanks.

vaultah commented:
I believe delete files should be fixed separately, otherwise there'd be too many changes to validate. There are also several issues related to delete files that can be addressed together (https://github.com/apache/iceberg/issues/12554 + https://github.com/apache/iceberg/issues/12586)

vaultah commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -87,6 +87,24 @@ public Set<Pair<String, String>> copyPlan() {
     }
   }
 
+  public static class RewrittenFileInfo implements Serializable {
```
We could [return it from `rewrite*ManifestWithResult`](https://github.com/apache/iceberg/pull/13720#discussion_r2252819385) and maybe use it in fixes for delete files later

vaultah commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -491,77 +528,128 @@ public RewriteContentFileResult appendDeleteFile(RewriteResult<DeleteFile> r1) {
+    public Map<String, RewrittenFileInfo> getRewrittenManifests() {
+      return rewrittenManifests;
+    }
+  }
```
I don't see a way to do that because `RewriteContentFileResult` in its current form is not needed for position deletes. It should be okay to have a small wrapper class just for manifests rewrite result. I have now made it private though.

vaultah commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -357,6 +438,55 @@ public static RewriteResult<DataFile> rewriteDataManifest(
+                        writer))
+            .reduce(new RewriteResult<>(), RewriteResult::append);
+    }
+    return Pair.of(writer.toManifestFile(), rewriteResult);
```
It now returns the size of the rewritten manifest instead of the full instance.

vaultah commented:
Also it seems problematic that IO exceptions are skipped [here](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java#L284-L286). If it's to handle expired snapshots with missing manifest lists, that method is only supposed to be called with live snapshots, and also `ManifestList.read` appears to throw `NotFoundException` for missing files so it's not caught anyway

dramaticlly commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -1209,6 +1213,191 @@ public void testNestedDirectoryStructurePreservation() throws Exception {
+
+    // Rewrite table metadata to a location that's much longer than the original in order
+    // to make manifests larger
+    String targetLocation = toAbsolute(rootTargetLocation) + generateLongNestedPath(25);
```
I think you can actually just move this into the function above and no need to keep `@TempDir Path rootTargetLocation` for each tests

```java
  protected String targetTableLocation() {
    return toAbsolute(targetTableDir) + generateLongNestedPath(5);
  }
```

dramaticlly replied:
I tried with longer targetTableLocation for all tests and they seem to be happy 😉  But I am fine with `longTargetTableLocation`

dramaticlly commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -1209,6 +1213,191 @@ public void testNestedDirectoryStructurePreservation() throws Exception {
+
+    Map<String, Long> manifestSizesBeforeRewrite =
+        sourceTable.currentSnapshot().allManifests(sourceTable.io()).stream()
+            .collect(Collectors.toMap(m -> fileName(m.path()), m -> m.length()));
```
nit: not relate to your change, but I just realized for filename(m.path()), we can just replace with just `Paths.get(path).getFileName().toString()` and and `m.length()` can be replaced with Method reference

dramaticlly commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -1209,6 +1213,191 @@ public void testNestedDirectoryStructurePreservation() throws Exception {
+    Snapshot lastSnapshot = targetTable.currentSnapshot();
+
+    List<ManifestFile> allManifests = lastSnapshot.allManifests(targetTable.io());
+    ManifestFile rewrittenManifest =
+        Iterables.getOnlyElement(
+            allManifests.stream()
+                .filter(manifest -> manifest.snapshotId() == lastSnapshot.snapshotId())
+                .collect(Collectors.toList()));
```
looks like this `rewrittenManifest` is not used, do we want to apply some assertion to this or can be removed?

vaultah replied:
Removed it.

dramaticlly commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -1209,6 +1213,191 @@ public void testNestedDirectoryStructurePreservation() throws Exception {
+    return pathBuilder.toString();
+  }
+
+  protected void addAnyPositionDelete(Table targetTable, String path) throws Exception {
```
nit: how about

`protected void commitNewPositionDelete(Table table, String deleteFilePath)`

dramaticlly reviewed (commented):
Overall LGTM, some nit comments.

vaultah commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -335,7 +393,9 @@ public static RewriteResult<DataFile> rewriteDataManifest(
    * @return a copy plan of content files in the manifest that was rewritten
+   * @deprecated since 1.10.0, will be removed in 1.11.0
    */
+  @Deprecated
```
That method is being used by `RewriteTablePathSparkAction` in Spark 3.4 and Spark 3.5, so we can't remove it unless we backport these changes to Spark 3.4 and Spark 3.5 in the same (this) PR. Same for `rewriteDataManifest` and `rewriteDeleteManifest` methods below. Let me know how you want to proceed

stevenzwu replied:
if this PR is going to take longer due to larger refactoring for the suggested bottom up approach and result classes and can't make the 1.10 release in time, we can go with the deprecation approach like here. Let's defer this deprecation decision /change as the last step.

dramaticlly commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -447,6 +551,58 @@ public static RewriteResult<DeleteFile> rewriteDeleteManifest(
+   *     rewritten here)
+   * @return size of the resulting manifest file and a copy plan for the referenced content files
+   */
+  public static Pair<Long, RewriteResult<DeleteFile>> rewriteDeleteManifestWithResult(
```
I think there's value in this, as in this PR we are doing bottom first rewrite in order to have size, this differs from previous top down approach and result might need to change accordingly. We return `Pair<Long, RewriteResult<DeleteFile>>` when rewrite a result of delete manifests for both long and discovered list of deletes to copy and to rewrite, all of which belong to a single delete manifests. 

For other top down rewrite such as given metadata.json to discover list of delta snapshots, size is unknown and we can leave it null. It's can be up to caller to leverage the size in the rewrite result. Let's give it a try and see if there's compatibility problem

vaultah replied:
I understand and agree that it makes sense to add size attribute to rewrite result, but the way `RewriteResult` is currently implemented and used is counterintuitive to me. When I see `RewriteResult<T>`, I expect the result of rewriting `T`. Adding another general-purpose and nullable attribute that's not related to `T` makes it even more confusing. I did consider doing exactly that initially, but it felt like shoehorning, attempting to attach metadata about the container (the manifest file's size) to an object that describes its contents. Same for `RewriteContentFileResult`, which is public too.

That said, if we're aligned that modifying `RewriteResult` is the preferred path forward, I'm happy to implement it that way.

dramaticlly replied:
I understand your confusion and I think we can probably do better on this RewriteResult<T> class. It's counterintuitive in a way that the instance of this result is not just for rewrite of T, but a collection T require open/rewrite/close in next iteration and also a collection of file path to copy from source to destination. 

Take manifestList as an example, itself is a avro files contain mapping to many manifest list. So when we trying to path rewrite a manifestList, it contains list of manifests require open to rewrite, list of source to destination path to plan the copy. Conceptually I think it  such result is relate to a manifest-list, now we are adding the size of written manifest lists. Recursively this also applies to when we apply path rewrite on manifest level, results contain set of data files, list of copy plans and size of manifest. 

For existing rewriteManifestList method in utils, the return type is of RewriteResult<ManifestFile> instead of RewriteResult<ManifestListFile>. Ans similarly the rewriteVersionFiles are returning RewriteResult<Snapshot>.

dramaticlly commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -494,77 +478,128 @@ public RewriteContentFileResult appendDeleteFile(RewriteResult<DeleteFile> r1) {
-          break;
+                  targetPrefix);
+          return Tuple3.apply(
+              manifestFile.path(),
```
I think some of the complexity of this can contribute to the fact that the size is treated outside the RewriteResult, so we are bundling tuple3 by ManifestFile in MappingFunction and unzip & reassembly in localIterator with necessary aggregation. 

Can we explore on bundle the size in RewriteResult and see `Map<String, RewriteContentFileResult>` helps with such situation?

vaultah commented:
- Added nullable `size` to `RewriteResult`
 - Removed `ManifestRewriteResult`
 - Kept `RewriteContentFileResult` unchanged (public and can't be made private)
 - Removed `rewrite*ManifestWithResult` methods
 - Removed deprecation notice from  `rewrite*Manifest` methods that are now enough
 
PTAL

vaultah commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -281,24 +281,32 @@ private String rebuildMetadata() {
         Sets.difference(snapshotSet(endMetadata), snapshotSet(startMetadata));
 
+    // rebuild manifest files
+    Set<ManifestFile> manifestsToRewrite = manifestsToRewrite(validSnapshots);
```
> manifest files from all live snapshots

Manifest files from live snapshots between (`startVersion`, `endVersion`].

Snapshot rewrite remains incremental. We have to rewrite all manifests _referenced_ in the selected live snapshots, to get their sizes after the rewrite. We cannot update sizes only for manifests _added_ in these snapshots, because that would leave other manifests in manifest lists with incorrect sizes.

stevenzwu replied:
thanks both for explaining the difference to me

vaultah commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -417,55 +424,37 @@ private Set<Pair<String, String>> statsFileCopyPlan(
 
-  private Set<String> manifestsToRewrite(
-      Set<Snapshot> deltaSnapshots, TableMetadata startMetadata) {
+  private Set<ManifestFile> manifestsToRewrite(Set<Snapshot> liveSnapshots) {
```
IMO yes, it returns the manifests that we need to rewrite.

> it returns manifest files from all live snapshots

Manifests referenced by live snapshots between `startVersion` and `endVersion`.

vaultah commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -276,6 +287,62 @@ public static RewriteResult<ManifestFile> rewriteManifestList(
+   * @param targetPrefix target prefix that will replace it
+   * @param outputPath location to write the manifest list
+   */
+  public static void rewriteManifestList(
```
I think this method is too low-level for `RewriteTablePathSparkAction` and moving it there would scatter logic even more, also it uses `ManifestLists` that's not public in `org.apache.iceberg` and can't be imported to `RewriteTablePathSparkAction` that's in `org.apache.iceberg.spark.actions`. We could move most of `RewriteTablePathUtil.rewriteManifestList` logic and only keep  a small wrapper method around `ManifestLists.write`, or make `ManifestLists` public. Let me know how you'd like to proceed

vaultah commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -347,16 +414,24 @@ public static RewriteResult<DataFile> rewriteDataManifest(
+              .reduce(new RewriteResult<>(), RewriteResult::append);
     }
+
+    rewriteResult.setSize(writer.length());
```
I don't think either is ideal, but I changed it to `length`.

stevenzwu replied:
+1 on length

vaultah commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -281,24 +281,32 @@ private String rebuildMetadata() {
     // rebuild position delete files
     Set<DeleteFile> deleteFiles =
-        rewriteManifestResult.toRewrite().stream()
+        allManifestsResult.toRewrite().stream()
```
I don't think readability gains are substantial enough to justify double-iteration, but I implemented your version.

stevenzwu commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -494,36 +483,60 @@ public RewriteContentFileResult appendDeleteFile(RewriteResult<DeleteFile> r1) {
+                    sourcePrefix,
+                    targetPrefix),
+                tupleEncoder)
+            .toLocalIterator();
```
In the previous code of using `reduce`, I thought the executors perform the initial reduction within their partitions, then the driver aggregates the partial results from executors.

With `toLocalIterator`, everything is shipped back to the driver for one pass of aggregation. Hence I was asking about the memory footprint and scalability for large tables.

vaultah replied:
We have to build a map of manifests and memory footprint should be acceptable because `toLocalIterator` fetches data to the driver one partition at a time. We can probably still use reduce to pre-aggregate maps on executors and merge them into one big map on the driver if you think that'd be preferable.

huaxingao replied:
I also think this may cause problem. With `toLocalIterator`, every record is shipped to the driver and aggregated there in one pass. It streams partition-by-partition, but the driver still processes O(N) records and the accumulator can grow to O(#manifests). For large tables that can increase driver memory and hurt scalability.

stevenzwu replied:
can we still implement the aggregation using the `reduce` pattern here (commutative and associative )?

stevenzwu replied:
The complexity comes from the awkward design of the `RewriteResult` and `RewriteContentFileResult`.

I would hope that we can do a larger refactoring to simplify the whole thing. There could be a `RewriteManifestResult` which internally tracks the original manifest path along with rewrite results.

There is no need to track `toRewrite` if we are switching to the bottom up approach. There is no need for `copyPlan` that is needed for top-down. We just need the original file path and the result. The result may contain a few attributes: new path, new size etc.

vaultah replied:
That's a great point. I agree that a larger refactoring to create a dedicated, reducible result class would be a much cleaner design in the long run.

Given that this PR is already quite large and focused on the correctness bug, would you be open to tackling that refactoring in a follow-up PR? I can create a new issue to track that work.

Does that sound like a reasonable plan?

dramaticlly replied:
Sounds fair to me and I think it helps to have correctness fix first

huaxingao replied:
Agree. Let’s proceed with the minimal correctness fix now. Please open a tracking issue and add a TODO with a link in the code.

stevenzwu commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -281,24 +281,34 @@ private String rebuildMetadata() {
     // rebuild position delete files
     Set<DeleteFile> deleteFiles =
-        rewriteManifestResult.toRewrite().stream()
+        allManifestsResult.toRewrite().stream()
```
nit: is the variable renaming more accurate? trying to see if it is unnecessary change.

dramaticlly commented on .palantir/revapi.yml:
```diff
@@ -1262,6 +1262,12 @@ acceptedBreaks:
+        \ java.lang.Object>"
+      new: "class org.apache.iceberg.RewriteTablePathUtil.RewriteResult<T extends\
+        \ java.lang.Object>"
+      justification: "Adding an optional field"
```
I think "Serialization across versions is not supported" is better and conform to result of `java.class.defaultSerializationChanged`

dramaticlly reviewed (approved):
LGTM and pending CI, thanks @vaultah for the contribution

dramaticlly commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -230,7 +230,8 @@ public void testStartVersion() throws Exception {
 
-    checkFileNum(1, 1, 1, 4, result);
+    // 1 metadata JSON file, 1 snapshot, 2 manifests, 1 data file
+    checkFileNum(1, 1, 2, 5, result);
```
@stevenzwu raised a good point on this length fix shall not change the behavior for delta rewrite. I recall we had a earlier discussion on this in https://github.com/apache/iceberg/pull/13720#discussion_r2292103624 but on a second thought I believe we are trying to do many things together, to rewrite extra manifests which is not added by delta snapshot to fix the manifest size in new manifest-list. 

```mermaid
graph TD
  subgraph "Version File Metadata"
    V2["v2.json"]
    V3["v3.json"]
  end

  subgraph "Manifest Lists"
    S1["snap1.avro"]
    S2["snap2.avro"]
  end

  subgraph "Manifests"
    M1["m1.avro"]
    M2["m2.avro"]
  end


  V2 --> S1
  V3 --> S2
  S1 --> M1
  S2 --new--> M2
  S2 --existing--> M1
```
Let's use this unit test as an example here, if we do incremental rewrite from v2.metadata.json

- Before:
we scan manifests table to filter manifests where added_snapshot_id is in deltaSnapshotIdSet, so we will only need to rewrite 1 manifest (m2.avro)

- After:
we scan all_manifests table to filter manifests where reference_snapshot_id is in deltaSnapshotIdSet, so we will have to rewrite 2 manifests (m1.avro & m2.avro)

I think this might help with fix the previous manifest size problem in the new manifest-list but we are no longer doing incremental path rewrite (where only added snapshot shall be able to determine the delta).

dramaticlly reviewed (changes requested):
I think we want to keep the original delta rewrite behavior before we can merge this

vaultah commented:
@dramaticlly @stevenzwu 

Let's say we have manifests A, B, and C, added by snapshot 1. We then create snapshot 2: its manifest list will contain references to manifests A, B, and C, and we also add the reference to the new manifest D. 

Now we use `rewrite_table_path` in incremental mode, starting from snapshot 1. 

Per your suggestion, it will rewrite just manifest D and the manifest list of snapshot 2. We assume that manifests added in snapshot 1 were already rewritten before, so we simply update their paths in the manifest list of snapshot 2.

For manifest D, the rewritten manifest list will have the new path and the new size. For manifests A, B, C, the rewritten manifest list will have new paths and their original lengths. In other words, the rewritten manifest list will be

| manifest_path | manifest_length | ... |
| --- | --- | --- |
| newPath(D) | newLength(D)| |
| newPath(C) | oldLength(C)| |
| newPath(B) | oldLength(B)| |
| newPath(A) | oldLength(A)| |


As a result of rewriting, manifest length will almost certainly change, so in general `oldLength(A) != newLength(A)`, which means the size of manifest A in the rewritten manifest list is incorrect, as in it doesn't match the length of the actual physical file at `newPath(A)` that it's referencing. This is the scenario from https://github.com/apache/iceberg/issues/13719 that I'm trying to solve.

Please help me understand how the correctness is maintained in your suggestion

dramaticlly commented:
> @dramaticlly sorry, I'm not sure I follow. Re. **1**: My point is that even in incremental mode, even when we rewrite a single manifest list (snapshot 2), we need to update lengths for all manifests that it references (A, B, C, D). If we don't do that, per your suggestion, the rewritten manifest list will only have correct lengths for the manifests that were rewritten in the same run (D), leaving lengths of other manifests (A, B, C) as they were in the source table (i.e., _as if_ they were never rewritten) and therefore incorrect. Then, after this manifest list is added to the target table, the target table will have the problem described in #13719.
> 
> Re. **2**: fixing existing manifest references is not in scope of this PR. This PR fixes `rewrite_table_path` so it produces correct manifest lists going forward. In a correct manifest list, `manifest_length` matches the size of the file at `manifest_path` for all manifest references. To guarantee that, we need to rewrite all referenced manifests.

Re 1. I get what you mean now, even though manifest A, B, C was copied earlier with correct length in target table, the incremental rewrite for snapshot 2 cannot obtain such length from source table, so they unfortunately need rewrite to get the correct length. And this essentially means incremental rewrite is no longer limited to the changing snapshots. Let me think a bit more on this.

vaultah commented:
Hi @dramaticlly, just wanted to check in on this. Please let me know if you have any thoughts on the path forward when you get a chance. Thanks!

dramaticlly commented:
> Hi @dramaticlly, just wanted to check in on this. Please let me know if you have any thoughts on the path forward when you get a chance. Thanks!

I am going to raise this in iceberg community sync tmr (Sept 3rd 9am PT) https://iceberg.apache.org/community/#apache-iceberg-community-calendar and collect some feedback. Feel free to join us if time works for you
